File: server/pkg/model/message.go
package model

type ContentType int

const (
	ContentText          ContentType = 0
	ContentImage         ContentType = 1
	ContentFile          ContentType = 2
	ContentAudio         ContentType = 3
	ContentVideo         ContentType = 4
	ContentSystem        ContentType = 5
	ContentRecall        ContentType = 6
	ContentEdit          ContentType = 7
	ContentCustom        ContentType = 8
	ContentAgentTimeline ContentType = 9
	ContentForm          ContentType = 10
	ContentFormResponse  ContentType = 11
)

type MsgStatus int

const (
	MsgSending   MsgStatus = 0
	MsgSent      MsgStatus = 1
	MsgDelivered MsgStatus = 2
	MsgRead      MsgStatus = 3
)

type Message struct {
	MsgID           int64       `json:"msg_id"`
	ConvID          string      `json:"conv_id"`
	SenderID        string      `json:"sender_id"`
	SenderName      string      `json:"sender_name"`
	SenderSessionID string      `json:"sender_session_id"`
	ContentType     ContentType `json:"content_type"`
	Body            string      `json:"body"`
	Mention         []string    `json:"mention,omitempty"`
	ReplyTo         int64       `json:"reply_to"`
	Timestamp       int64       `json:"timestamp"`
	ClientSeq       int64       `json:"client_seq"`
	ConvSeq         int64       `json:"conv_seq"`
	Status          MsgStatus   `json:"status"`
	Deleted         bool        `json:"deleted,omitempty"`
}

File: server/web/src/types/message.ts
export enum ContentType {
  Text = 0,
  Image = 1,
  File = 2,
  Audio = 3,
  Video = 4,
  System = 5,
  Recall = 6,
  Edit = 7,
  Custom = 8,
  AgentTimeline = 9,
  Form = 10,
  FormResponse = 11,
}

export enum MsgStatus {
  Sending = 0,
  Sent = 1,
  Delivered = 2,
  Read = 3,
}

export interface Message {
  msg_id: number
  conv_id: string
  sender_id: string
  sender_name: string
  sender_session_id: string
  content_type: ContentType
  body: string
  mention?: string[]
  reply_to: number
  timestamp: number
  client_seq: number
  conv_seq: number
  status: MsgStatus
  deleted?: boolean
}

// For local messages before server assigns msg_id
export interface LocalMessage extends Message {
  stableId: string
}

export function messageStableId(msg: Message): string {
  return msg.msg_id > 0 ? String(msg.msg_id) : `local-${msg.client_seq}`
}
